src/csharp/Grpc.Core/Interceptors/ClientInterceptor.cs

```diff
@@ -0,0 +1,281 @@
+            where TResponse : class
+        {
+            return next(method, host, options);
+        }
+
+        /// <summary>
+        /// Intercepts an asynchronous invocation of a duplex streaming call.
+        /// </summary>
+        public virtual AsyncDuplexStreamingCall<TRequest, TResponse> AsyncDuplexStreamingCall<TRequest, TResponse>(Method<TRequest, TResponse> method, string host, CallOptions options, Func<Method<TRequest, TResponse>, string, CallOptions, AsyncDuplexStreamingCall<TRequest, TResponse>> next)
+            where TRequest : class
+            where TResponse : class
+        {
+            return next(method, host, options);
+        }
+
+        /// <summary>
+        /// Returns a new instance of <c>AsyncUnaryCall</c> that has hooks to intercept on an underlying object of the same type.
+        /// </summary>
+        protected static AsyncUnaryCall<TResponse> Intercept<TRequest, TResponse>(AsyncUnaryCall<TResponse> call,
+            Func<Task<TResponse>, TResponse> response = null,
+            Func<Task<Metadata>, Metadata> responseHeaders = null,
+            Func<Func<Status>, Func<Status>> getStatus = null,
+            Func<Func<Metadata>, Func<Metadata>> getTrailers = null,
+            Func<Action, Action> dispose = null)
+        {
+            GrpcPreconditions.CheckNotNull(call, "call");
+            var callResponse = call.ResponseAsync;
+            Task<Metadata> callResponseHeaders = call.ResponseHeadersAsync;
+            Func<Status> callStatus = call.GetStatus;
+            Func<Metadata> callTrailers = call.GetTrailers;
+            Action callDispose = call.Dispose;
+
+            if (callResponse != null)
+            {
+                var responseOriginal = callResponse;
+                callResponse = new Task<TResponse>(() => response(responseOriginal));
+            }
+
+            if (responseHeaders != null)
+            {
+                var callResponseHeadersOriginal = callResponseHeaders; // Ensure it will not be captured by the closure
+                callResponseHeaders = new Task<Metadata>(() => responseHeaders(callResponseHeadersOriginal));
+            }
+
+            if (getStatus != null)
+            {
+                callStatus = getStatus(callStatus);
+            }
+
+            if (getTrailers != null)
+            {
+                callTrailers = getTrailers(callTrailers);
+            }
+
+            if (dispose != null)
+            {
+                callDispose = dispose(callDispose);
+            }
+
+            return new AsyncUnaryCall<TResponse>(callResponse, callResponseHeaders, callStatus, callTrailers, callDispose);
+        }
+
+        /// <summary>
+        /// Returns a new instance of <c>AsyncServerStreamingCall</c> that has hooks to intercept on an underlying object of the same type.
+        /// </summary>
+        protected static AsyncServerStreamingCall<TResponse> Intercept<TRequest, TResponse>(AsyncServerStreamingCall<TResponse> call,
```

Comment:
They are provided for interceptor implementors to be able to intercept additional events in async or streaming calls. The initial interceptor hook is only invoked at the beginning of the call.

These functions are only needed because the constructor for `AsyncXYZCall` are internal. If we are going to make them public, now would be a good time. I could just remove all of this mess.

Reply:
I'd quite like a way of creating an interceptor to just modify/inspect one aspect (e.g. metadata) of all calls - `Interceptor.ForMetadata(metadata => { ...})`

Reply:
I agree, and have a few other helper extension methods in mind. I wanted to keep this PR minimal to the core hooks to make it easier to review... now I am thinking it might have been best to add such things for additional context?